cmake/SpectreCheckTargetDependencies.cmake

```diff
@@ -0,0 +1,667 @@
+    list(APPEND SPECTRE_INCLUDES ${INCLUDE})
+  endforeach(INCLUDE ${_RAW_SPECTRE_INCLUDES})
+  # "Return" SPECTRE_INCLUDES by setting it in PARENT_SCOPE
+  set(SPECTRE_INCLUDES ${SPECTRE_INCLUDES} PARENT_SCOPE)
+endfunction(_extract_spectre_includes FILE_NAME)
+
+# Get the list of 3rd party library includes for the target `TARGET_NAME`.
+#
+# Sets the variable `TPL_INCLUDES` in parent scope to return the result
+function(_extract_tpl_includes FILE_NAME)
+  file(READ ${FILE_NAME} FILE_CONTENTS)
+  string(REGEX MATCHALL "\n#include <[^\n]+>"
+    _RAW_TPL_INCLUDES ${FILE_CONTENTS})
+  unset(TPL_INCLUDES)
+  foreach(INCLUDE ${_RAW_TPL_INCLUDES})
+    string(REPLACE "\n#include <" "" INCLUDE "${INCLUDE}")
+    string(REPLACE ">" "" INCLUDE "${INCLUDE}")
+    list(APPEND TPL_INCLUDES ${INCLUDE})
+  endforeach(INCLUDE ${_RAW_TPL_INCLUDES})
+  # "Return" TPL_INCLUDES by setting it in PARENT_SCOPE
+  set(TPL_INCLUDES ${TPL_INCLUDES} PARENT_SCOPE)
+endfunction(_extract_tpl_includes FILE_NAME)
+
+# Make the path of `HEADER_FILE` which belongs to the target `TARGET_NAME`
+# an absolute path.
+#
+# Sets the variable `ABS_HEADER_FILE` in parent scope to return the result
+function(_make_header_path_absolute TARGET_NAME HEADER_FILE)
+  get_property(
+    _INCLUDE_DIR
+    TARGET ${TARGET_NAME}
+    PROPERTY INTERFACE_INCLUDE_DIRECTORIES
+    )
+  if(IS_ABSOLUTE ${HEADER_FILE})
+    set(_ABS_HEADER_FILE ${HEADER_FILE})
+  else()
+    set(_ABS_HEADER_FILE "${_INCLUDE_DIR}/${HEADER_FILE}")
+  endif()
+  # "Return" ABS_HEADER_FILE by setting it in PARENT_SCOPE
+  set(ABS_HEADER_FILE ${_ABS_HEADER_FILE} PARENT_SCOPE)
+endfunction(_make_header_path_absolute TARGET_NAME HEADER_FILE)
+
+# Gets the dependencies for a list of includes of one of the source or
+# header files of the target `TARGET_NAME`
+#
+# The `_get_deps_for_includes` function builds a list of all dependencies
+# by finding the target that provides each header file in `INCLUDES_OF_FILE`.
+# `INCLUDES_OF_FILE` must be a list of includes from a file (either header
+# or source file) of the target `TARGET_NAME`. If `RECURSE_THROUGH_INCLUDES`
+# is `TRUE` then for each header file in `INCLUDES_OF_FILE` that is also an
+# include of the target `TARGET_NAME` then the header file will also have its
+# dependencies searched and added to the list of dependencies.
+#
+# Example usage:
+#
+#   _get_deps_for_includes(
+#     TARGET_C
+#     "TARGET_A;TARGET_B;TARGET_C;TARGET_D"
+#     "cmath;utility;vector"
+#     FALSE
+#     )
+#
+# Arguments:
+#
+# TARGET_NAME: the target for which to check the link libraries/dependencies
+# LIST_OF_ALL_TARGETS: a list of all the targets that might be dependencies
+#                      of the target ${TARGET_NAME}
+# INCLUDES_OF_FILE:
+#     a list of header files for which to get the dependencies.
+# RECURSE_THROUGH_INCLUDES:
+#     if TRUE then any include in any of ${INCLUDES_OF_FILE}
+#     that is also in the list ${TARGET_HEADER_FILES} has its dependencies
```

Comment:
`TARGET_HEADER_FILES` is an internal variable of the function so it might be nicer to say what it means in terms of the input arguments...